File: 0234-palindrome-linked-list/0234-palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if (!head || !head.next) return true
    let slow = head
    let fast = head
    let prev
    while (fast && fast.next) {
        prev = slow
        slow = slow.next
        fast = fast.next.next
    }
    prev.next = null
    prev = null
    while (slow) {
        const temp = slow.next
        slow.next = prev
        prev = slow
        slow = temp
    }
    slow = head
    fast = prev
    while (slow && fast) {
        if (slow.val !== fast.val) return false
        slow = slow.next
        fast = fast.next
    }
    return true
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head:
            return True
        
        def reverse(prev: Optional[ListNode], head: Optional[ListNode]) -> Optional[ListNode]:
            if not head:
                return prev
            next = None
            if head.next:
                next = ListNode(head.next.val, head.next.next)
            head.next = prev
            return reverse(head, next)
        
        reversed = reverse(None, ListNode(head.val, head.next))
        a, b = reversed, head
        while a and b:
            if a.val != b.val:
                return False
            a = a.next
            b = b.next
        if a or b:
            return False
        else:
            return True